File: engines/lua/g2048/hash.lua
-- SPEC §7: FNV-1a 32-bit board and history hashes.

local M = {}

local MASK = 0xffffffff
local OFFSET = 0x811C9DC5
local PRIME = 0x01000193

--- fnv1a32 over the bytes of a string.
function M.fnv1a32(s)
  local h = OFFSET
  for i = 1, #s do
    h = ((h ~ s:byte(i)) * PRIME) & MASK
  end
  return h
end

--- fnv1a32 over the 16 exponent bytes.
function M.board_hash(b)
  local h = OFFSET
  for i = 1, 16 do
    h = ((h ~ b[i]) * PRIME) & MASK
  end
  return h
end

--- fnv1a32(le32(h) ++ 16 board bytes ++ [dir]).
function M.history_step(h, b, dir)
  local x = OFFSET
  for k = 0, 3 do
    x = ((x ~ ((h >> (8 * k)) & 0xff)) * PRIME) & MASK
  end
  for i = 1, 16 do
    x = ((x ~ b[i]) * PRIME) & MASK
  end
  return ((x ~ dir) * PRIME) & MASK
end

--- 8 lowercase hex digits.
function M.hex(h)
  return string.format("%08x", h)
end

return M

File: engines/java/src/main/java/dev/g2048/engine/Hash.java
package dev.g2048.engine;

/** FNV-1a state hashes (SPEC §7). */
public final class Hash {
  private Hash() {}

  private static final int OFFSET = 0x811C9DC5;
  private static final int PRIME = 0x01000193;

  public static int fnv1a32(byte[] data) {
    int h = OFFSET;
    for (byte c : data) h = (h ^ (c & 0xff)) * PRIME;
    return h;
  }

  public static int boardHash(byte[] b) {
    int h = OFFSET;
    for (int i = 0; i < 16; i++) h = (h ^ (b[i] & 0xff)) * PRIME;
    return h;
  }

  /** fnv1a32(le32(h) ++ board ++ [dir]). */
  public static int historyStep(int h, byte[] b, int dir) {
    int x = OFFSET;
    for (int k = 0; k < 4; k++) x = (x ^ ((h >>> (8 * k)) & 0xff)) * PRIME;
    for (int i = 0; i < 16; i++) x = (x ^ (b[i] & 0xff)) * PRIME;
    return (x ^ dir) * PRIME;
  }

  /** 8 lowercase hex digits. */
  public static String hex(int h) {
    String s = Integer.toHexString(h);
    return "00000000".substring(s.length()) + s;
  }
}
